### ISSUE TITLE
 Add ElasticSearch 5.0 support

### ISSUE THREAD

**Issue by JanMalte** [django-haystack/django-haystack#1383]
elasticsearch 5.0 is in alpha already and should be added on future release of haystack. Maybe [skipping ealsticsearch 2.0 support](https://github.com/django-haystack/django-haystack/issues/1247) would be an option. [elasticsearch-py](https://github.com/elastic/elasticsearch-py) already supports elasticsearch 5.0

https://www.elastic.co/guide/en/elasticsearch/reference/master/breaking-changes-5.0.html

**Comment by kcolton** [comment]
FWIW, I think ElasticSearch 2.0 is still an important initiative. As references Elastic 5.0 has [breaking changes](https://www.elastic.co/guide/en/elasticsearch/reference/master/breaking-changes.html) and upgrade path looks a little bumpy, so I think adoption will take time.

**Comment by mahmoudhossam** [comment]
ES 5.0 final has been released.

**Comment by SalahAdDin** [comment]
What's about 2.x.x?

**Comment by kcolton** [comment]
@SalahAdDin I think @mahmoudhossam was referring to the fact that ElasticSearch 5.0 is beyond Release Candidate and is now the current _stable_ officially supported ElasticSearch release. 

There are some currently open Pull Requests #1391 #1336  but AFAIK there is nothing beyond that in terms of official Haystack support for ES 2.x.x

The best thing to do may be to try running off of those branches and seeing if you bump into issues along the way. That may be the best insight into how close those planes are to landing

---

I think this brings up an interesting discussion point about some core architectural decisions within Haystack. Here's an interesting question:

**What is the least common denominator of functionality across the search backends?**

In the earlier days, where each backend was more or less simple document search with some different takes on implementation. Basic faceting and field support was also pretty common. So the decisions in that context made sense.

Now though, the backends breadth of features are diverging and specializing even more. The best common functionality I can see is just simple document search. And even that is not as backend agnostic as it sounds. (Differences in analyzer capabilities etc)

This is in contrast to RDBMSs which has SQL, which although "standard" is quite loose - the base set of features and functions a SQL interface supports is much larger and makes something like a backend agnostic and robust `QuerySet` possible. For example - implementations as different as `SQLite` and `Redshift` have a large set of common features executed with a common enough interface that a large portion of django ORM functions just work.

Given that each backend seems to be diverging and specializing, I've started to wonder if it might make sense to _reduce_ the set of functions that come with your vanilla `SearchQuerySet`? 

This is just as a hypothetical extreme - but imagine if vanilla "core" haystack just did full document search, no fields, no facets, no stats, no spatial search, nothing much beyond `.filter(content=)` with a small set of `InputType`s. It would still be hugely powerful and much easier to keep the "core" up to date across backends.

Then "contrib" packages come in to extend with more backend specific specific features and  functionality. 

This seems like it could make it vastly easier to maintain the core (which does benefit from staying up to date with engines) and for those specialists using the larger set of features for each backend to maintain the backend-specific extensions to `SearchQuerySet` and `SearchIndex`.

I know this is a bit of a radical idea. Penny for your thoughts @acdha ?

**Comment by mahmoudhossam** [comment]
I agree with @kcolton, haystack needs to be modular to gain traction.

It's been hard to push a new release of haystack because each backend needs to be updated, and that is a monumental effort for a single software package.

Splitting haystack into core and a bunch of plugins or contrib packages would really help the project.

**Comment by acdha** [comment]
This is a long discussion and I don't have enough time to do it justice
right now but in general I agree. Haystack's design reflects decisions made
about the state of Django and Python packaging close to a decade ago and
it'd be good to overhaul it significantly for a Haystack 3 release. The
idea which has come up in the past – I think @honzakral made a good
proposal a couple years back – would be simply to embrace an extension
point to really simplify getting a connection handle for the appropriate
backend client so once you hit the limits of what the core SearchQuerySet
can do you have a clean way to do something along the lines of what
Django's raw/extra mechanism allows, only more-so since the functionality
diverges more widely. Even core things like just searching for records has
a pretty wide range with things like Solr's different query parsers and the
way ElasticSearch allows you to either do basic Solr-like queries or the
rich JSON interface which is really powerful and thus even less likely to
be abstractable in a portable manner.

I've made a few improvements in this regard – e.g. my backend supporting
the Solr collapse/expand grouping system (
https://gist.github.com/acdha/0a66ca23984bc8d607936fecd9c29941) required
less boilerplate than the older one I developed against Haystack 2.0 in the
beta era (https://gist.github.com/acdha/3750774) – but I really think that
would be a good area for further improvement so we could just embrace the
pattern of making it easy to recognize when you need to jettison
portability anyway and get out of the way.

One initial step I'd like to do is identify a maintainer for each major
backend and either refactor entirely into separate apps
(django-haystack-core, -solr, -es, -es2, -whoosh, etc.) or at least put
some time into further reducing the amount of time which someone who is
interested in contributing in one backend needs to know about the others
for non-core features. I use Solr primarily and use ES casually but it'd be
really nice if we could delegate most of the support for each backend to
someone who actively tracks the upstream community and would be able to
maintain a richer .contrib module without having to take on the problem of
building a portable abstraction.

Chris

**Comment by HonzaKral** [comment]
I am sorry, but I do not have the bandwidth to support elasticsearch backend for haystack. I decided some time ago to instead make an elasticsearch-specific library [0] that should hopefully replace it for many use cases. It is not as tightly integrated with django but supports the whole rang of options elasticsearch has to offer.

Big parts of it could also be used by haystack to abstract away some of the differences between elastic versions (`filtered` query vs `{"bool": {"filter": []}}` for example) and make the overall haystack code a bit simpler. It even provides hooks for custom (de)serialization and other goodies. I would also be happy to add more if it would make haystack easier to port.

Feel free to open any issues there or get in touch directly with me, I am always happy to help pythonistas with elasticsearch issues.

Thank you

0 - https://github.com/elastic/elasticsearch-dsl-py

**Comment by kcolton** [comment]
@acdha @HonzaKral Firstly, thanks both of you! 

I think Haystacks _current modularly_ has tremendously helped in gaining traction. But as @acdha pointed out, the world of search backends is moving rapidly with the ecosystem looking very different 3 years ago.

We've been using Elastic since 0.X for over 3 years and a variety of use cases. We've basically tried every combination of Elastic + Django.

- Mozilla's deprecated [elasticutils](https://github.com/mozilla/elasticutils)
- just using `elasticsearch-dsl-py` and essentially @HonzaKral's example of [integrating with django](https://github.com/HonzaKral/es-django-example)
- server side querying through `elasticsearch-py`, `elasticsearch-dsl-py`, direct client side querying which predated the JS DSL (dark times)

Where we have landed is actually a combination of both `haystack`with a few extensions specifically designed to make use of the lovely [elasticsearch-dsl-py](https://github.com/elastic/elasticsearch-dsl-py).

Our approach has been very much in line with the [Solr collapse/expand grouping system](https://gist.github.com/acdha/0a66ca23984bc8d607936fecd9c29941) example but for adding Elastic specific functionality like nested aggregations, pre-search filters that don't go into the query string, etc. As mentioned, we use `elasticsearch-dsl-py`to keep the interface neat. 

Rather than trying to "fix" `.facet`, we just made a `.aggregate` method that takes in either dictionary definition (if you like pain) or DSL objects and a `.pre_filter` on an extended version of `ElasticSearchQueryset`: https://gist.github.com/kcolton/d143503fd25a694ffce4cd7e178349ae

Upgrading to 2.X and beyond aside - we have definitely ran into some challenges that seem to come from complexities that _other_ `SearchQuerySet` methods introduce and large functions tat could be broken down for more extensibility or additional hook points. That's what seeded the idea that maybe the vanilla `SearchQuerySet` getting smaller would not be a bad thing.

Haystack + DSL hybrid, for our use case optimizes for least number of wheel reinventions. Going just DSL meant having to worry about: 

- Django object "rehydration" from search results (with efficient prefetching)
- Signal processing (we use `haystack-celery` + more extensions for related object support + random batch commands)
- An interface that is familiar for Django developers and fits (very efficiently with a few more extensions) into Djangos existing generic views

We, [The Material World Team](https://angel.co/material-wrld#team) are committed to staying on Haystack _and_ to getting to 5.0 in the next few months. We are happy to contribute any way we can as we owe much to both Haystack (ty again @acdha + contributors) and the DSL for python (ty again @HonzaKral and crew)

Getting the functionality running in for a private project with specific use cases is one thing, and we are happy to contribute what and where we can, but would most likely need more community support for staying on top of ES specific features - particularly the ones that fall out of our use cases.

Need to catch up on the current 2.X efforts (it seems like some of them are close?) and possibly put together a proposal for a "path to 5.0" and how it might look with and without 2.X support.

Cheers,

Ken Colton
CTO Material World

**Comment by barseghyanartur** [comment]
@kcolton:

And you're right to use `elasticsearch-py` and `elasticsearch-dsl-py`!

This project is as good as dead now. I don't know what could bring it back to live (kick-starter or any other campaign), but nowadays almost nobody uses SOLR or Whoosh with Python. But there's nobody around to take the problem (lack of modern ElasticSearch support) seriously. Therefore, die peacefully, haystack. You've done a good job in past, but you have become irrelevant.

**Comment by NDevox** [comment]
For those interested I ported the current elasticsearch engine over to 2.X. I haven't managed to get it working with 5.x yet.

You can find it here: https://pypi.python.org/pypi/elasticsearch2-haystack

It should work for all versions up to 2.4 but it isn't perfect as it essentially patches 2.x to look like 1.x.

From what I've heard there is someone working on a newer backend which is hopefully optimised for the latest releases of ES which this isn't as far as I'm concerned. So I'm not going to make a PR - I'm just making it available until there is a better replacement.

We've tested it at our work on Django-Oscar and it works just fine right now.

**Comment by Alkalit** [comment]
I currently work on elasticsearch 5 backend for haystack. In my opinion it, currently, would be better to have are separate backends for each version.

**Comment by NDevox** [comment]
@Alkalit. I think this is definitely the way to go. We just needed a backend which supported 2.x quickly.

**Comment by acdha** [comment]
@Alkalit Agreed – I think the history of breaking changes is likely to continue and trying to maintain backwards compatibility in the same code-base has definitely made maintaining Haystack expensive in the past.

**Comment by kcolton** [comment]
@acdha The more I've been thinking about it - the "least common denominator" between  what one would consider a search backend (beyond the 4 currently) seems to basically only be ... "search":

### Least Common Denominator of all "search backends"

- full text document search with a relevancy score (although not necessarily boosting). 
  - Basically a tautology. Otherwise, it wouldn't be a search index. In contrast to a traditional RDBMS that had no fulltext search capability and "search" meant `WHERE foo LIKE '%?%' OR bar LIKE '%?%' OR baz LIKE '%?%'`
- pagination 

### Where Haystack Fits In

Easily connecting scalable, flexible and robust document based, full-text search to Django models and perform a generic search across them is _extremely powerful_. 

This was the mental elevator pitch for us regarding `haystack`: Real search for our Django project in minutes that just works and won't be a maintenance headache. 

Plenty of modern RDBMS's have damn good full-text field search, but the fact that so many projects are still using haystack or other search solutions show that the being able to search across models and customize documents is important.

Aside from document generation Haystack does some things very well that are wheels anyone making a search framework would have to invent:

- Keeping the index up to date. A search index is basically a cache and [you know they say about caching](http://martinfowler.com/bliki/TwoHardThings.html). The signal processing framework, which allows for an indexing strategy that meets many business requirements. For us, and likely many simple web applications - the combination of `celery-haystack` with its post commit hooks +  periodic randomized re-indexing to correct any "drift" is an indexing strategy that is fast yet fault tolerant.

- A Django friendly, familiar interface for fetching data for views that includes efficient "rehydration" of objects from search results. (We actually stick with normal forms and generic CBVs rather than use haystacks)

Think of all the additional backends (search platforms + search as a service) that could easily be connected and still get the value of the above 2 properties.

Our use case actually falls outside the 90% "just need document search", as we _do_ require the more esoteric features of ElasticSearch, yet at the moment haystack still strikes a good balance and has saved us many hours of wheel re-invention.

@acdha This is an extreme, but interesting thought experiment:

Image where `SearchIndex` only facilitates rendering a document, signal processing, efficient and safe reindexing and rehydration. **No fields**

Where the `SearchQuerySet` vanilla interface is `models()`, `.auto_query()`, `__len__()`, `__iter__()`, `__getitem__()` + model rehydration. **No filter, no exclude, no facet, etc**

Basically - a `django-haystack-lite`. Normally, rewrites are doomed to failure, but why I think it's more "haystack-lite" than "haystack 3.0" is that it's mostly _removal_ of code.

I know that really takes things to an extreme and is not practical - but I'm struggling to think of a way to architect something that would be able to keep up with the seemingly exponential changes in the search world at large that isn't as extreme. 

Particularly the fact that most search backends are no longer just search. A single search backend like ElasticSearch can be used for NoSQL OLTP, OLAP, data warehousing, big data crunching, realtime communication, glued into ETL pipelines, and many other use cases that have almost nothing to do with search. Many of them would inevitably become feature requests and the current foundation makes extension to support such features difficult. 

I actually can't even imagine _one_ `ElasticSearchBackend` anymore without it turning into a kitchen sink itself. There's enough different fairly exclusive use cases within that 1 platform to be several backends.

This is a tough problem man. Once again, penny for your (and all others) thoughts?

Cheers,
Ken

**Comment by NDevox** [comment]
@kcolton I more or less agree with all of that. Features should be available on a backend by backend basis and we shouldn't aim for everything in all backends.

It would make more sense to follow an ORM approach in my mind as well - where each engine is a separate project instead of tying them in to the core or Haystack.

**Comment by SalahAdDin** [comment]
Please, support it!

**Comment by snirp** [comment]
What is the current status regarding ES5 support? Seems like this issue has gotten quiet without a game plan.

**Comment by tuxedojoe** [comment]
As a newcomer to Django and search overall, Haystack was a wonderfull way to start. I do not think Haystack's days are counted. On the opposite. The possibility to use a Django plugin that I can use locally and try out with Whoosh, Solr and then finally Elasticsearch was an amazing experience.

Search is commonly a small add-on, and the ability to just get up and running with the possibility of scaling up the search chain (simple > whoosh > elasticsearch) is just the most common use-case for adding search in my small experience. So Haystack do really play a big and important role for all Django and future Django developers.

So I think ElasticSearch 5 support is gonna get important, so I upvote this post! And if there is anyway we, the newcomers can help in that, I would love to give that a try. However, I start with an upvote.

**Comment by barseghyanartur** [comment]
@tuxedojoe:
It won't happen any time soon with or without up votes. If you need latest elastic, don't use haystack. Don't assume support for 5.0 will be out in less than two years.

**Comment by SalahAdDin** [comment]
@barseghyanartur What other alternative we can use?

**Comment by acdha** [comment]
If you're interested in actually contributing to Haystack, please test this branch and report your results. I don't use ES much and thus far we haven't had much luck getting volunteers to help with support for it.

**Comment by barseghyanartur** [comment]
@SalahAdDin:

I'm quite happy with elasticsearch-dsl and django-elasticsearch-dsl.

**Comment by SalahAdDin** [comment]
@barseghyanartur , what's about help here?

**Comment by barseghyanartur** [comment]
@SalahAdDin:

Regarding the "help here":

- https://github.com/django-haystack/django-haystack/pulls/barseghyanartur
- https://github.com/django-haystack/django-haystack/commits?author=barseghyanartur

Additionally, see my GitHub account. I contributed to a lot of open-source projects. 

In general, regarding the subject of "Haystack" or "no-Haystack". I just want a tool that "works" and works well. Who uses SOLR, Whoosh, Xapian nowadays? ElasticSearch is used much broader. 

The Haystack API is very limited. If you want to do some serious work with ElasticSearch, you will find yourself fighting against Haystack, having a need to constantly hacking, overriding, sub-classing, etc. Code that you finally get looks dirty. FYI, I did use Haystack with ElasticSearch 2.x, I know what I'm talking about.

ElasticSearch evolves very fast. Haystack evolves extremely slow. Do you want to put all Django developers who are making their first steps with ElasticSearch in trouble? Because they are going to be in trouble if they start with Haystack, as a year later product owner will say he wants to use features of ElasticSearch 5.0 and Haystack hardly supports the 2.0 branch. 

Besides, if you have chosen ElasticSearch, you're unlikely willing to migrate to SOLR, Xapian or any other backend Haystack has support for. Would migrate to SQLite from PostgreSQL?

**Comment by SalahAdDin** [comment]
@barseghyanartur What's about projects like `django-oscar` that use `Haystack` for multiple backend support?

I think, you are right, but, in my case, for `django-oscar` there isn't support for these projects. In the case of `Wagtail` projects, they did like you, a little, they implemented his own `elasticsearch` backend and finish.

**Comment by jonesnc** [comment]
@acdha silly question: how do I test the code related to this issue? Which branch should I test?

**Comment by acdha** [comment]
@jonesnc Yes - I believe it's at the point where the test suite should pass but could use more review, especially just for basic things like documentation / setup instructions. One of the other things I need to do is see whether landing #1504 broke anything.

**Comment by jonesnc** [comment]
I'm not sure if this has been discussed elsewhere, but [elasticsearch 5.x no longer supports the `string` datatype](https://www.elastic.co/guide/en/elasticsearch/reference/current/string.html).

As it stands now, https://github.com/django-haystack/django-haystack/pull/1461 still uses `field_type = 'string'` for `indexes.CharField` (among other classes), which results in `WARN` log messages that look like:

> [WARN ][o.e.d.i.m.StringFieldMapper$TypeParser] The [string] field is deprecated, please use [text] or [keyword] instead on [text]

**Comment by jonesnc** [comment]
Is there any work still being done on this issue?

What still needs to be done, and how can I help?

**Comment by pmaccamp** [comment]
@jonesnc field_type = 'string' is breaking if elasticsearch cannot perform automatic upgrades due to other parameters so would like to see it addressed in the PR

`
elasticsearch.exceptions.RequestError ... The [string] type is removed in 5.0 and automatic upgrade failed because parameters [term_vector] are not supported for automatic upgrades. You should now use either a [text] or [keyword] field instead ...
`

**Comment by jonesnc** [comment]
@pmaccamp And is this the most up-to-date fork which addresses this issue?

https://github.com/CraveFood/django-haystack

**Comment by pmaccamp** [comment]
@jonesnc yes that fork is still using ` field_type='string' `

I'm also running into another issue when querying `no [query] registered for [filtered]` that appears be to due to changes how filtered queries are constructed with ES 5 https://stackoverflow.com/questions/40519806/no-query-registered-for-filtered.

**Comment by machakux** [comment]
Elasticsearch is evolving very fast  and has a lot of features. It seems to me it won't be easy for the whole `django-haystack` to catch up on Elasticsearch soon.
There are other solutions for using utilizing Elasticsearch in Django including Elasicsearch DSL and its derivatives but I think haystack still provides a nicer SearchQuerySet syntax.

There are few custom haystack backends for Elasticsearch which may be found by searching on Internet but I haven't managed to find one which is actively mantained and supports Elasticsearch 5. Therefore  I created a separate project for pluging in Elasticsearch 5 support and other Elasticsearch-specific features to Haystack.

The source code can be found at https://github.com/tehamalab/django-haystack-es . So far it seems to work in indexing, faceting, boosting and few other things but I haven't done much testing and documentation. I hope it will be helpful in one way or another.

**Comment by kcolton** [comment]
I still think the `ExtremelyBaseSearchQuery` and `ExtremelyBaseSearchBackend` ([Approach referenced last year](https://github.com/django-haystack/django-haystack/issues/1383#issuecomment-266317855)) is a solid approach.

* It to keeps the core of haystack minimal, well supported, and kept up to date with Django, especially with DJ 2.0 on the horizon
* Makes it easier to implement new backends as they come out and removes some of the complexities that make the existing `BaseSearchQuery` and `BaseSearchBackend` difficult to extend.
* Can open up to the explosion of new search backends and services

@acdha  Wondering if you had much of a pulse on the relative usage of each of the features.  Would be really interesting to run a poll like:

* What backend(s) are you running?
* Do you use index fields?
* Simple faceted fields?
* Date faceted fields?
* Distance fields?
* Signal processors
* Realtime signal processors
* Celery signal processors
* The view, form and template helpers
* Have you extended or customized components? Which?

etc 

Having the base engine class have to support this just seems untenable:

```python
def build_search_kwargs(self, query_string, sort_by=None, start_offset=0, end_offset=None,
                            fields='', highlight=False, facets=None,
                            date_facets=None, query_facets=None,
                            narrow_queries=None, spelling_query=None,
                            within=None, dwithin=None, distance_point=None,
                            models=None, limit_to_registered_models=None,
                            result_class=None):
```

Every search engine is specializing in one way or another, this interface just seems too wide now. Even within elasticsearch you might have a backend that cares a lot about distance, or one that does nested objects, or one that focuses on analytics. 

I have a fork as a POC for this idea and it seemed to work well. It created a new base class (in spirit, i narrowed a interfaces of a subclass just to avoid conflicts for the POC) and then setup new backends and query's based on those and got it to a point where all tests at the time were running again.

I need to pull it up to date, it's a year behind, but i'll put out a PR of it out as an RFC to get some more brains on the idea.

**Comment by codekiln** [comment]
[Elasticsearch 6.0 is now out](https://www.elastic.co/guide/en/elasticsearch/reference/6.0/release-notes-6.0.0.html)

**Comment by SalahAdDin** [comment]
@codekiln, Holly shit!

**Comment by codekiln** [comment]
@all, what are your thoughts on "[django-haystack-elasticsearch](https://pypi.python.org/pypi/django-haystack-elasticsearch)"? It's listed as Pre-Alpha and hasn't (as of the time of this writing) been updated on PyPy since last year. Are people using it successfully with ES 5 and vanilla haystack, or does it go with CraveFood's fork of haystack?

**Comment by jonesnc** [comment]
@codekiln By `PyPy`, I think you mean `[PyPI](https://pypi.python.org/pypi)`

**Comment by SalahAdDin** [comment]
@codekiln, are you sure about it?

**Comment by codekiln** [comment]
@jonesnc you're right, I mean PyPI.
@SalahAdDin what do you mean? sure about what? elasticsearch 6.0 is definitely out

**Comment by SalahAdDin** [comment]
@codekiln, these package works with `elasticsearch 6.0`?

**Comment by codekiln** [comment]
@SalahAdDin I'm asking, actually. I'm considering using raw elasticsearch-dsl-py and elasticsearch-py, but wanted to see if anyone had used or does use "django-haystack-elasticsearch".

**Comment by acdha** [comment]
@codekiln How tightly are you planning to follow ElasticSearch? If you don't need portability, I would consider going with the native elasticsearch packages since they pay developers to work on them. Haystack's selling point is portability but that's also somewhat expensive in terms of the time required to track something like ES which changes interfaces regularly, especially since we haven't had a whole lot of community interest in that work.

**Comment by acdha** [comment]
I just rebased and merged #1461
